### DISCUSSION TITLE
`EquationBC` doesn't seem to work in adjoint optimization problems

### DISCUSSION THREAD

**Discussion by vojtamolda** [firedrakeproject/firedrake#2774] [Firedrake support] [answered]
Hello everybody,

First of all, thanks for the work and effort you put into this framework. Especially the adjoint feature feels almost like a magical piece of technology and I have very much enjoyed solving PDEs with it.

I'm working on a topological optimization problem that under the hood uses the `firedrake_adjoint` module to calculate gradients of the objective function w.r.t to some parametrization. I formulated my PDEs using the `EquationBC` boundary condition; but it seems like it isn't supported in adjoint enabled of problems. So my question is whether this is a valid conclusion or if there is an error between the chair and the keyboard.

Below is a toy Poisson problem that works fine in "pure" Firedrake:

```python
from firedrake import *
# from firedrake_adjoint import *  # if uncommented, instantiation of `EquationBC` fails

mesh_num = 16
mesh = UnitSquareMesh(mesh_num, mesh_num)

V = FunctionSpace(mesh, "CG", 1)

u = TrialFunction(V)
v = TestFunction(V)

f = Function(V)
x, y = SpatialCoordinate(mesh)
f.interpolate(- 8.0 * pi * pi * cos(x * pi * 2) * cos(y * pi * 2))

a = - inner(grad(u), grad(v)) * dx
L = inner(f, v) * dx

g = Function(V)
g.interpolate(cos(2 * pi * x) * cos(2 * pi * y))

u_ = Function(V)

bc1 = EquationBC(inner(u, v) * ds(1) == inner(g, v) * ds(1), u_, 1)

solve(a == L, u_, bcs=[bc1])

f.interpolate(cos(x * pi * 2) * cos(y * pi * 2))
cost = sqrt(assemble(inner(u_ - f, u_ - f) * dx))
```

Once the `firedrake_adjoint` module gets imported, I can't find a way to instantiate the `EquationBC` object. Any attempt so far failed with:

```
AttributeError: 'EquationBC' object has no attribute 'block_variable'
```

I noticed that the implementation of `EquationBC` compared to [`DirichletBC`](https://github.com/firedrakeproject/firedrake/blob/14fe574486578183adb3bcf04f7d83b64d51fe63/firedrake/bcs.py#L441) is missing a lot of the `*_ad_annotate_*` decorators. I'm not sure what these exactly do but it may very well be the case that nobody actually needed to use the `EquationBC` in an adjoint problem before.

There's a relatively simple workaround in my case - to add Neumann/Robin BC terms to the weak form of the PDE. So the problem isn't a deal breaker; I just want to satisfy my intelectual curiosity and find out the answer.

**Comment by ksagiyam** [comment, marked as answer]
You are right that `EquationBC` currently does not work with adjoint. There is a work-in-progress PR on this (https://github.com/firedrakeproject/firedrake/pull/2036), but we have yet to do some work to make this fully functional.
Thanks!

**Reply by vojtamolda** [reply]
Thanks for the quick reply! I'll check out the PR.
